sasl/sasl.go

```diff
@@ -42,3 +44,21 @@ type StateMachine interface {
 	// value will be true.
 	Next(ctx context.Context, challenge []byte) (done bool, response []byte, err error)
 }
+
+// Metadata contains additional data for performing SASL authentication.
+type Metadata struct {
+	Address string
+}
+
+// WithContext returns a copy of the context with associated Metadata.
+func (m *Metadata) WithContext(ctx context.Context) context.Context {
```

Comment:
What made you choose a method over a top-level function here?

Reply:
I like the direct approach of embedding itself into a context given a parent (examples of the same approach out in the wild https://github.com/rs/zerolog/blob/master/ctx.go#L28)

Reply:
I think `WithContext` is a bit weird as a method name. For example, `http.(*Request).WithContext` returns a `*http.Request` which embeds the context, `context.WithValue` returns a context which embeds a value, etc...

Here we say "metadata with context" but don't return a metadata embedding a context, we return a context which embeds the metadata that it was called on... effectively inverting the pattern established by the standard library.

In my opinion, the top-level function seems to be the most in line with the way the standard library is constructed:
```go
func WithMetadata(ctx context.Context, m *Metadata) context.Context { ... }
```